src/devices/bus/rs232/sync_io.cpp

```diff
@@ -0,0 +1,213 @@
+	PORT_RS232_BAUD("RS232_BAUD" , RS232_BAUD_2400 , "Baud rate" , sync_io_device , update_serial)
+
+	PORT_START("rtsduplex")
+	PORT_CONFNAME(1 , 1 , "RTS controls half-duplex")
+	PORT_CONFSETTING(0 , DEF_STR(Off))
+	PORT_CONFSETTING(1 , DEF_STR(On))
+
+	PORT_START("txcontrol")
+	PORT_CONFNAME(1 , 1 , "TxC generated")
+	PORT_CONFSETTING(0 , "Always")
+	PORT_CONFSETTING(1 , "When Tx is active")
+
+	PORT_START("rxcontrol")
+	PORT_CONFNAME(3 , 1 , "RxC generated")
+	PORT_CONFSETTING(0 , "Always")
+	PORT_CONFSETTING(1 , "When Rx is active")
+	PORT_CONFSETTING(2 , "When a byte is available")
+INPUT_PORTS_END
+
+ioport_constructor sync_io_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME(sync_io);
+}
+
+void sync_io_device::device_add_mconfig(machine_config &config)
+{
+	BITBANGER(config , m_stream , 0);
+}
+
+void sync_io_device::device_start()
+{
+	m_clk_timer = timer_alloc(TMR_ID_CLK);
+}
+
+void sync_io_device::device_reset()
+{
+	output_dcd(0);
+	output_dsr(0);
+	output_rxd(1);
+	output_cts(1);
+
+	update_serial(0);
+
+	m_clk = false;
+	m_rx_byte = 0xff;
+	m_rx_counter = 0;
+	m_tx_counter = 0;
+	m_tx_enabled = false;
+	m_rx_enabled = false;
+
+	output_txc(1);
+	output_rxc(1);
+}
+
+void sync_io_device::device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr)
+{
+	switch (id) {
+	case TMR_ID_CLK:
+		m_clk = !m_clk;
+		if (m_clk) {
+			output_txc(1);
+			output_rxc(1);
+			if (m_tx_enabled) {
+				// Rising edge: capture TxD
+				m_tx_byte = (m_tx_byte >> 1) & 0x7f;
+				if (m_txd) {
+					BIT_SET(m_tx_byte , 7);
+				}
+				if (++m_tx_counter >= 8) {
+					LOG("Tx %02x @%s\n" , m_tx_byte , machine().time().to_string());
+					m_tx_counter = 0;
+					m_stream->output(m_tx_byte);
+					m_tx_enabled = false;
+				}
+			}
+		} else {
+			if (!m_tx_enabled) {
+				m_tx_enabled = !m_rts || !m_rts_duplex->read();
+				output_cts(!m_tx_enabled);
+			}
+			if (m_tx_enabled || m_txc_setting->read() == 0) {
+				output_txc(0);
+			}
+			if (!m_rx_enabled) {
+				m_rx_enabled = m_rts || !m_rts_duplex->read();
+				if (m_rx_enabled) {
+					if (m_stream->input(&m_rx_byte , 1) == 0) {
+						m_rx_byte = ~0;
+						if (m_rxc_setting->read() == 2) {
+							m_rx_enabled = false;
+						}
+					} else {
+						LOG("Rx %02x @%s\n" , m_rx_byte , machine().time().to_string());
+					}
+				}
+			}
+			if (m_rx_enabled || m_rxc_setting->read() == 0) {
+				output_rxc(0);
+			}
+			if (m_rx_enabled) {
+				// Falling edge: update RxD
+				output_rxd(BIT(m_rx_byte , 0));
+				m_rx_byte >>= 1;
+				if (++m_rx_counter >= 8) {
+					m_rx_counter = 0;
+					m_rx_enabled = false;
+				}
+			}
+		}
+		break;
+
+	default:
+		break;
+	}
+}
+
+WRITE_LINE_MEMBER(sync_io_device::update_serial)
+{
+	auto baud_rate = convert_baud(m_rs232_baud->read());
+	auto period = attotime::from_hz(baud_rate * 2);
+	m_clk_timer->adjust(period , 0 , period);
+}
+
+DEFINE_DEVICE_TYPE(SYNC_IO, sync_io_device, "sync_io", "RS232 Synchronous I/O")
```

Comment:
The class and device names are too generic.  Please use something more specific.